cloudify_agent/operations.py

```diff
@@ -446,3 +451,62 @@ def validate_agent_amqp(current_amqp=True, **_):
         raise NonRecoverableError(result['agent_alive_error'])
     if not current_amqp and not result['agent_alive_crossbroker']:
         raise NonRecoverableError(result['agent_alive_crossbroker_error'])
+
+
+@operation
+def transfer_agent_amqp(transfer_agent_timeout=300,
+                        manager_ip=None, manager_certificate=None,
+                        manager_rest_token=None, **_):
+
+    _create_broker_config(transfer_mode=True)
+    old_agent = _validate_agent()
+    agents = _run_install_script(old_agent, transfer_agent_timeout, manager_ip,
+                                 manager_certificate, manager_rest_token,
+                                 transfer_agent=True)
+    returned_agent = agents['new']
+    ctx.logger.info('Configured agent {0} to work with the new Manager'.
+                    format(returned_agent['name']))
+
+    # Make sure the agent is alive:
+    app = get_celery_app(tenant=returned_agent['rest_tenant'])
+    _assert_agent_alive(returned_agent['name'], app)
+
+
+def _create_broker_config(transfer_mode=False):
+    """
+    This function creates a dictionary called 'broker_config' within the
+    'cloudify_agent' dict, that contains all the required information that will
+     later be used to create a Celery client
+    :return:
+    """
+    if 'cloudify_agent' not in ctx.instance.runtime_properties:
+        raise NonRecoverableError(
+            'cloudify_agent key not available in runtime_properties')
+    agent = ctx.instance.runtime_properties['cloudify_agent']
+    agent['broker_config'] = dict()
+    broker_conf = agent['broker_config']
+    broker_conf['broker_ip'] = agent.get('broker_ip')
+    broker_conf['broker_user'] =\
+        (agent.get('rest_tenant')).get('rabbitmq_username')
+    broker_conf['broker_vhost'] =\
+        (agent.get('rest_tenant')).get('rabbitmq_vhost')
+    broker_conf['broker_pass'] =\
+        (agent.get('rest_tenant')).get('rabbitmq_password')
+    broker_conf['broker_ssl_enabled'] = True
+    if transfer_mode:
+        status = ctx.instance.runtime_properties['agent_status']
+        status['agent_alive_crossbroker'] = True
+        ssl_path = agent.get('broker_ssl_cert_path')
+        with open(ssl_path, 'r') as ssl_file:
+            ssl_cert = ssl_file.read()
+        broker_conf['broker_ssl_cert'] = ssl_cert
+
+
+def _conceal_amqp_password(url):
+    """
+    replace the broker password in the url before printing it
+    """
+    no_psw = url[:url.find(':', 5)] + url[url.find('@'):]
```

Comment:
The usual approach is to break the string into 3 parts: before password, password, after password. Then join them into before password, ***, after password